libfwupdplugin/fu-bios-attrs.c

```diff
@@ -0,0 +1,334 @@
+		return;
+	}
+	if (!fu_strtoull(scalar, &tmp, 0, G_MAXUINT64, &error_local)) {
+		g_debug("%s", error_local->message);
+		return;
+	}
+	fwupd_bios_attr_set_lower_bound(attr, tmp);
+}
+
+static void
+fu_bios_attr_set_type(FuBiosAttrs *self, FwupdBiosAttr *attr, const gchar *driver)
+{
+	g_autofree gchar *data = NULL;
+	g_autoptr(GError) error_local = NULL;
+
+	g_return_if_fail(FU_IS_BIOS_ATTRS(self));
+	g_return_if_fail(FWUPD_IS_BIOS_ATTR(attr));
+	g_return_if_fail(driver != NULL);
+
+	if (g_strcmp0(fwupd_bios_attr_get_name(attr), FWUPD_BIOS_ATTR_PENDING_REBOOT) == 0) {
+		fwupd_bios_attr_set_kernel_type(attr, FWUPD_BIOS_ATTR_TYPE_READ_ONLY);
+		return;
+	}
+
+	/* lenovo thinklmi seems to be missing it even though it's mandatory :/ */
+	if (!fu_bios_attr_get_key(attr, "type", data, &error_local)) {
+		g_debug("%s", error_local->message);
+		if (!self->kernel_bug_shown)
+			g_warning("KERNEL BUG: %s doesn't export a 'type' attribute", driver);
+		self->kernel_bug_shown = TRUE;
+		fwupd_bios_attr_set_kernel_type(attr, FWUPD_BIOS_ATTR_TYPE_ENUMERATION);
+		return;
+	}
+
+	if (g_strcmp0(data, "enumeration") == 0) {
+		fwupd_bios_attr_set_kernel_type(attr, FWUPD_BIOS_ATTR_TYPE_ENUMERATION);
+		return;
+	}
+	if (g_strcmp0(data, "integer") == 0) {
+		fwupd_bios_attr_set_kernel_type(attr, FWUPD_BIOS_ATTR_TYPE_INTEGER);
+		fu_bios_attr_set_integer_attrs(attr);
+		return;
+	}
+	if (g_strcmp0(data, "string") == 0) {
+		fwupd_bios_attr_set_kernel_type(attr, FWUPD_BIOS_ATTR_TYPE_STRING);
+		return;
+	}
+}
+
+static void
+fu_bios_attrs_populate_attribute(FuBiosAttrs *self,
+				 const gchar *driver,
+				 const gchar *path,
+				 const gchar *name)
+{
+	g_autoptr(FwupdBiosAttr) attr = NULL;
+	g_autoptr(GError) error_local = NULL;
+
+	g_return_if_fail(FU_IS_BIOS_ATTRS(self));
+
+	attr = fwupd_bios_attr_new(name, path);
+	fu_bios_attr_set_type(self, attr, driver);
+	fu_bios_attr_set_description(attr);
+	g_ptr_array_add(self->attrs, g_object_ref(attr));
+}
+
+/**
+ * fu_bios_attrs_reinitialize:
+ * @self: a #FuBiosAttrs
+ *
+ * Clears all attributes and re-initializes them.
+ * Mostly used for the test suite, but could potentially be connected to udev
+ * events for drivers being loaded or unloaded too.
+ *
+ * Since: 1.8.4
+ **/
+void
+fu_bios_attrs_reinitialize(FuBiosAttrs *self)
+{
+	g_return_if_fail(FU_IS_BIOS_ATTRS(self));
+	g_ptr_array_set_size(self->attrs, 0);
+	fu_bios_attrs_init(self);
+}
+
+static void
+fu_bios_attrs_init(FuBiosAttrs *self)
+{
+	g_autofree gchar *sysfsfwdir = NULL;
+	g_autofree gchar *tmp = NULL;
+	g_autoptr(GError) error_local = NULL;
+	const gchar *name;
+	const gchar *driver;
+	g_autoptr(GDir) class_dir = NULL;
+
+	g_return_if_fail(FU_IS_BIOS_ATTRS(self));
+
+	self->attrs = g_ptr_array_new_with_free_func((GDestroyNotify)g_object_unref);
+
+	sysfsfwdir = fu_path_from_kind(FU_PATH_KIND_SYSFSDIR_FW_ATTRIB);
+	class_dir = g_dir_open(sysfsfwdir, 0, &error_local);
+	if (class_dir == NULL) {
+		g_debug("%s", error_local->message);
```

Comment:
i really don't like the local printing; can we return a GError and FALSE

Reply:
fixed